torch/csrc/jit/serialization/type_importer.cpp

```diff
@@ -0,0 +1,162 @@
+#include <torch/csrc/jit/serialization/type_importer.h>
+
+#include <torch/custom_class.h>
+
+namespace torch {
+namespace jit {
+TypePtr TypeImporter::import(const TypePtr& origType_) {
+  auto it = remappedTypes_.find(origType_);
+  if (it != remappedTypes_.end()) {
+    return it->second;
+  }
+  // This is a hack to make sure we properly re-map function types. Otherwise,
+  // if we have two distinct TypePtrs that hold the same Function*, we will
+  // remap them independently (leading to the creation of two remapped
+  // Functions). See [remapping followups].
+  if (auto origFunctionType = origType_->cast<FunctionType>()) {
+    auto it = std::find_if(
+        remappedTypes_.cbegin(), remappedTypes_.cend(), [&](const auto& pr) {
+          if (auto fn = pr.first->template cast<FunctionType>()) {
+            return origFunctionType->function() == fn->function();
+          }
+          return false;
+        });
+    if (it != remappedTypes_.cend()) {
+      return it->second;
+    }
+  }
+  auto origType = origType_->cast<c10::NamedType>();
+  if (origType == nullptr || !origType->name()) {
+    // need to rewrite contained maps?
+    if (origType_->containedTypes().size()) {
+      auto newContained = fmap(
+          origType_->containedTypes(),
+          [&](const TypePtr& t) { return import(t); });
+      auto newType = origType_->withContained(std::move(newContained));
+      remappedTypes_.emplace(origType_, newType);
+      return newType;
+    } else {
+      // no need to remap this type
+      remappedTypes_.emplace(origType_, origType_);
+      return origType_;
+    }
+  }
+  if (getCustomClass(origType->name()->qualifiedName())) {
+    // We don't need to remap custom classes, because they are not managed by
+    // compilation units
+    remappedTypes_.emplace(origType_, origType_);
+    return origType_;
+  }
+  if (auto classType = origType->cast<ClassType>()) {
+    return remap(classType);
+  } else if (auto functionType = origType->cast<FunctionType>()) {
+    return remap(functionType);
+  } else if (auto interfaceType = origType->cast<InterfaceType>()) {
+    return remap(interfaceType);
+  } else if (auto tupleType = origType->cast<TupleType>()) {
+    // FIXME this is broken. For NamedTuples we need to perform the same
+    // remapping as everywhere else. But it substantially adds to the
+    // complexity of the remapping logic, since NamedTuples are structurally
+    // typed, unlike everything else.
+    //
+    // There is an additional unimplemented thing, where we need to remap the
+    // type of IValues that are burned into the graph as constants.
```

Comment:
Not just constants, but any type that appears in the graph that references existing types. For instance what happens to the graph node: `TupleCreate(ClassObject)`. The result type is not correctly mapped as far as I can tell.